### PR TITLE
WIP for continuous mode

### PR DESCRIPTION
**Pull request by adriaangraas**
WIP pull request for #4 for the purpose of feedback and testing.

### CODE REVIEW

File: include/slicerecon/reconstruction/reconstructor.hpp [hunk 1]
```diff
@@ -165,56 +174,86 @@ class reconstructor {
...
     slice_data reconstruct_slice(orientation x) {
+        // the lock is supposed to be always open if reconstruction mode == alternating
+        std::lock_guard<std::mutex> guard(gpu_mutex_);
+
         if (!initialized_) {
             return {{1, 1}, {0.0f}};
         }
 
-        return alg_->reconstruct_slice(x, 1 - write_index_);
+        return alg_->reconstruct_slice(x, write_index_);
```
Comment (jwbuurlage) [inline include/slicerecon/reconstruction/reconstructor.hpp]: Is this still correct for `alternating` mode, where we want to have one _write index_ and one _reconstruction index_?

File: src/util/processing.cpp [hunk 2]
```diff
@@ -24,9 +26,7 @@ void process_projection(bulk::world& world, int rows, int cols, float* data,
     int first_row = s * block_size;
     int final_row = std::min((s + 1) * block_size, rows);
 
-    int count = proj_id_max - proj_id_min + 1;
-
-    for (int proj = 0; proj < count; ++proj) {
+    for (int proj = proj_id_min; proj < proj_id_max + 1; ++proj) {
```
Comment (jwbuurlage) [inline src/util/processing.cpp]: I don't think I understand this change. The `data` buffer has size `count =  proj_id_max - proj_id_min + 1`. This is what was called the `group_size`; the number of projections that are buffered on the CPU. Did you change the buffer to be a full projection stack on the CPU as well?

File: src/reconstruction/reconstructor.cpp [hunk 3]
```diff
@@ -409,50 +422,100 @@ void reconstructor::initialize(acquisition::geometry geom) {
...
     environment_.spawn(parameters_.filter_cores, [&](auto& world) {
-        util::process_projection(
-            world, geom_.rows, geom_.cols, buffer_[write_index_].data(),
-            dark_.data(), flat_fielder_.data(), filter_, proj_id_min,
-            proj_id_max, !geom_.parallel, fdk_weights_,
-            !parameters_.already_linear, fft_plan_, ffti_plan_,
-            freq_buffer_[world.rank()], parameters_.retrieve_phase,
-            paganin_filter_, fft2d_plan_);
+        util::process_projection(world, geom_.rows, geom_.cols,
+                                 data, dark_.data(),
```
Comment (jwbuurlage) [inline src/reconstruction/reconstructor.cpp]: Here we put `buffer_[write_index_]` data to be processed. This data only has size equal to `group_size` or `update_every`, yet in the process function you seem to index with 'global (GPU) projection indices' (i.e. not within the group).

**Review by jwbuurlage — COMMENTED** [review]
I will take some time to review this. We should make sure this does not break the alternating mode.

- [ ] We should have some clear definitions of what buffers are for what, and what all the parameters mean. This will help to ensure that the functionality is 'correct'.
- [ ] Once we have this developer documentation, there should be also be documentation for the users for the new modes.
- [ ] Skimming through the code I saw some changes that may not be correct for alternating mode, I need to spend some time reading through the changes (see initial code comments).
- [ ] We need a CHANGELOG entry
- [ ] Maybe we should add an AUTHORS file to roughly document who provided what functionality.

File: include/slicerecon/util/data_types.hpp [hunk 4]
```diff
@@ -29,6 +29,15 @@ struct geometry {
 
 } // namespace acquisition
 
+/**
+ * The mode that the reconstructor is in. Continuous corresponds to a sliding-
+ * window. Alternating is the default.
+ * @see https://github.com/cicwi/SliceRecon/issues/4
+ */
+enum mode {
```
Comment (jwbuurlage) [inline include/slicerecon/util/data_types.hpp]: We should prefer scoped enumerations to allow 'alternating' and 'continuous' to be used as variable names, and to allow choosing the underlying type.

```cpp
enum class mode {
    alternating,
    continuous
};

// and

mode::alternating
mode::continuous
```

See also:
<https://en.cppreference.com/w/cpp/language/enum>

File: src/reconstruction/reconstructor.cpp [hunk 5]
```diff
@@ -409,50 +422,100 @@ void reconstructor::initialize(acquisition::geometry geom) {
...
     // For sinogram we want: [rows, projection_id, cols]
 
     for (int i = 0; i < geom_.rows; ++i) {
-        for (int j = 0; j < group_size; ++j) {
+        for (int j = proj_offset; j <= proj_end; ++j) {
             for (int k = 0; k < geom_.cols; ++k) {
-                sino_buffer[i * group_size * geom_.cols + j * geom_.cols + k] =
+                sino_buffer[i * buffer_size * geom_.cols + j * geom_.cols + k] =
                     projection_group[j * geom_.cols * geom_.rows +
                                      i * geom_.cols + k];
```
Comment (jwbuurlage) [inline src/reconstruction/reconstructor.cpp]: Here it also seems to me that `projection_group` (which is really `buffer_`) is being indexed with global projection indices while having a smaller size.

File: src/util/processing.cpp [hunk 6]
```diff
@@ -3,6 +3,8 @@
 #include <algorithm>
 #include <cmath>
 #include <complex>
+#include <unsupported/Eigen/FFT>
```
Comment (jwbuurlage) [inline src/util/processing.cpp]: We don't need this header anymore (switched to FFTW for 2D FFTs)

File: src/util/processing.cpp (hunk 2 above)
Comment (adriaangraas) [reply src/util/processing.cpp]: No, buffer on the CPU is of size `update_every` in continuous mode. But if `group_size` is smaller, then only part of the CPU buffer needs to be processed. Hence the loop should be over the group in the CPU buffer and not over [0, `count`[, where count equals the group size.

File: include/slicerecon/reconstruction/reconstructor.hpp (hunk 1 above)
Comment (adriaangraas) [reply include/slicerecon/reconstruction/reconstructor.hpp]: I have now substituted this in favor for `active_gpu_buffer_index_`.

File: src/util/processing.cpp (hunk 2 above)
Comment (adriaangraas) [reply src/util/processing.cpp]: Basically this does nothing more than that it allows the starting index in the data buffer to be different. Its necessary if you do not want to process the full `data`.

File: src/reconstruction/reconstructor.cpp (hunk 5 above)
Comment (adriaangraas) [reply src/reconstruction/reconstructor.cpp]: This is now always `update_every`, the exact size of the buffer. The sino buffer is also always of exactly size `update_every`.

File: src/reconstruction/reconstructor.cpp (hunk 3 above)
Comment (adriaangraas) [reply src/reconstruction/reconstructor.cpp]: Good catch, that was exactly where the edit in `processing.cpp` was for :)

**Comment by adriaangraas** [comment]
There was actually no need for a double `buffer_` variable. That, and for alternating mode I set the `update_every` to the `group_size`. This simplified code quite a bit as `write_index_` was not necessary to index `buffer_` anymore.

**Comment by jwbuurlage** [comment]
Thanks again for implementing this! Let me know when this is ready for a proper review.

Please see the TODO list above, and please also make sure that the old functionality still works the same :)

**Comment by adriaangraas** [comment]
In about two weeks I will properly test this, add some comments here and there and do the formal wrap up.

**Comment by jwbuurlage** [comment]
I have updated the processing system. After implementing Paganin filtering support, the processing function was becoming a mess. 

Instead, there is now a `ProjectionProcessor` which has a number of (optional) components. The parallelization is now over full projections, which should also make it easier for you to have an alternating/continuous mode by simply calling `projection_processor_.process` with the proper data position and projection count. The system is much better/flexible/modular this way.

I figured this was better to do pre-merge than post-merge. I don't think it should take too much time for you to adapt your changes to the new system.

File: include/slicerecon/util/data_types.hpp (hunk 4 above)
Comment (adriaangraas) [reply include/slicerecon/util/data_types.hpp]: Done.

**Comment by adriaangraas** [comment]
There are only three buffers now, the `buffer` for incoming data (processing is in-place there), the `sino_buffer` which is the transpose of the `buffer`, and the GPU buffer. Should be easy enough to document once we get there.

I also made a few changes to the code that made the difference between alternating and continuous mode more apparent, so that it is easier to understand and debug in the future.

I made a CHANGELOG entry, but I suggest that you make an AUTHORS file?

I've tested the different modes here and they seem to work alright. Would you like to give the code a try at your system?

File: CMakeLists.txt [hunk 7]
```diff
@@ -76,7 +76,7 @@ add_subdirectory(ext/tomopackets)
 add_subdirectory(${CMAKE_SOURCE_DIR}/ext/bulk)
 # --------------------------------------------------------------------------------------------
 # FFTW3
-find_package(FFTW REQUIRED)
+#find_package(FFTW REQUIRED)
```
Comment (jwbuurlage) [inline CMakeLists.txt]: The dependency on FFTW should be put in the CMake file, but I forgot to add a `FindFFTW.cmake` to the repository. This has now been pushed as a2ece54318892d2d643c48098241829078037b36. Can you test whether this now works on your system as well (and uncomment the `find_package` line)

File: src/util/processing.cpp [hunk 8]
```diff
@@ -3,6 +3,7 @@
 #include <algorithm>
 #include <cmath>
 #include <complex>
+#include <iostream>
```
Comment (jwbuurlage) [inline src/util/processing.cpp]: This is not used I think.

**Comment by jwbuurlage** [comment]
At first glance this looks very reasonable :+1: I will do a detailed review next week, and test it on my own system. Thanks!

**Comment by adriaangraas** [comment]
I think this is also the opportunity to change the behavior of `--group-size` in the alternating mode a bit.

Currently, if you specify group-size, this will override the size of `proj_count` from the geometry specification packet. If you send in 600 projections per rotation, and select a 300 group-size, then you will find your reconstruction rotated by 180 degrees every other rotation.

I propose to make `--group-size` purely the size of the group that is simultaneously processed. Then we have:
- `--group-size` the amount of simultaneous processed projections
- `--update-every` the size of the CPU buffer, and how often projections are uploaded to the GPU buffer
- `geom.proj_count` the size of the GPU buffer (of which we have two in alternating mode)

Then `--countinuous` writes in one buffer, and locks and alternating just has two buffers and switches writing to either one. The flag then does not alter any other functionality.

Currently in alternating mode, `--group-size` equals `--update-every`, but I think it'd be nice to be able to set the processing/upload batch size separately in this mode as well.

File: example/slicerecon_push_aps_32id.py [hunk 9]
```diff
@@ -95,7 +95,7 @@ def main(arg):
     binning = int(args.binning)
     subsampling = int(args.sample)  # taking 1 per (subsampling) instead of all the frames
 
-    nproj = 400  # number of projections per 180 degrees interval, this is coded
+    nproj = 600  # number of projections per 180 degrees interval, this is coded
```
Comment (jwbuurlage) [inline example/slicerecon_push_aps_32id.py:98]: Should we add this as a CLI argument?

File: example/slicerecon_push_aps_32id.py [hunk 10]
```diff
@@ -141,16 +141,16 @@ def main(arg):
         window_min_point = [-rx, -rx, -rz]  # x,y,z
         window_max_point = [rx, rx, rz]  # x,y,z
 
-        angles = np.linspace(0, np.pi, proj_count, endpoint=False) # np.mod(theta[0:nproj], np.pi)
+        angles = np.linspace(0, 2*np.pi, proj_count, endpoint=False) # np.mod(theta[0:nproj], np.pi)
```
Comment (jwbuurlage) [inline example/slicerecon_push_aps_32id.py:144]: I think it depends on if the parallel beam data set is 'continuous' (i.e. we have a mirrored set of projections, each other half-rotation) if we want to group 2 \pi together here. I think I also understand your comment "then you will find your reconstruction rotated by 180 degrees every other rotation" better now.

I don't know what acquisition is standard, we should think about how we want to define this.

Mathematically, parallel beam usually is defined on angles in [0, \pi], and this is the convention I use (implicitly) as well in SliceRecon. At this point, if you do continuous acquisition you should flip your (mirrored) projections accordingly.

**Comment by jwbuurlage** [comment]
I finally got around to reading through the code, nothing sticks out to me except for the minor comments I gave before.

I think there are still some things unclear about `group_size` and `update_every`.

In my original design:

- the _group size_ was about parallel computation: how coarsely do we want to perform our parallelism. I wanted at least a certain amount of data, so that each thread would have enough work to do.
- the parameter _update every_ was about refreshing reconstructions, regardless of computation. How often do we want to refresh the (outdated) slices and 3D preview.

In alternating mode, we want the reconstruction to build up instead of waiting for reconstruction until we have received the first full rotation (this is particularly relevant to slow lab setups like ours). After this, we want to refresh only after every full rotation. The way this worked is by updating the reconstruction every 'group_size (= update_every)' projections during the first rotation, and only every 'proj_count' in any subsequent rotation. I think this functionality would still make sense.

In continuous mode, I think the logical way is still to have 'group_size = update_every', but for all rotations, not only the first one.  So maybe, the extra flag 'update every' is actually not necessary.

For the parallel beam, I think your issue is that you should only send projections in the angular range [0, \pi]. See also my comment on the tomopy script.

File: example/slicerecon_push_aps_32id.py (hunk 10 above)
Comment (adriaangraas) [reply example/slicerecon_push_aps_32id.py:144]: This was #9 :)

File: example/slicerecon_push_aps_32id.py (hunk 9 above)
Comment (adriaangraas) [reply example/slicerecon_push_aps_32id.py:98]: Hm, good point. I'm afraid that the script is currently not generic enough to be used as a blackbox, and also don't want to leave that suggestion. Shall I rename it to `push_dynamic_foam` or something similar?

Comment (jwbuurlage) [reply example/slicerecon_push_aps_32id.py:98]: Yeah, with a link to the original dataset?

File: CMakeLists.txt (hunk 7 above)
Comment (adriaangraas) [reply CMakeLists.txt]: Cherry-picked that commit and works well!

**Comment by jwbuurlage** [comment]
Is this ready for a final check + merge?

**Comment by adriaangraas** [comment]
Do we want to remove the `update_every` from the argument list, and just automatically set it to the sensible defaults (i.e. equal to `group_size`)?

**Comment by jwbuurlage** [comment]
In my opinion, yes

**Comment by adriaangraas** [comment]
Okay, issue #9 was implicitly fixed in this branch and I've removed the `update_every` param. Things seem to run fine here, both with/without `--continuous` and `--group-size`. Would you like to make the final check before the merge?

**Comment by jwbuurlage** [comment]
Cool! Thanks.

**Pull request merged**